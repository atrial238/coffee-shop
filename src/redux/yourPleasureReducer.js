import {getCoffeePleasureAPI} from '../API/api';

const setIsError = (bool) => ({type: types.SET_IS_ERROR, body: bool}),
		setIsLoading = (bool) => ({type: types.SET_IS_LOADING, body: bool}),
		setPleasureCoffee = (data) => ({type: types.GET_PLEASURE_COFFEE, body: data});

export const types = {
	GET_PLEASURE_COFFEE: 'home_reducer/GET_PLEASURE_COFFEE',
	SET_IS_LOADING: 'home_reducer/SET_IS_LOADING',
	SET_IS_ERROR: 'home_reducer/SET_IS_ERROR',
} 


export const getCoffeePleasure = (methodAPI = getCoffeePleasureAPI) => (dispatch) =>  {
	dispatch(setIsError(false))
	dispatch(setIsLoading(true))
	return methodAPI()
		.then(res => {
			dispatch(setIsLoading(false))
			dispatch(setPleasureCoffee(res.coffeePleasure))
		})
		.catch(() => {
			dispatch(setIsLoading(false))
			dispatch(setIsError(true))
		})
}

const initState = {
	coffeePleasure: [],
	isLoading: false,
	isError: false
}

export const yourPleasureReducer = (state = initState, action) => {
	switch(action.type){
		case types.GET_PLEASURE_COFFEE:
			return {...state, coffeePleasure: action.body}
		case types.SET_IS_LOADING:
			return {...state, isLoading: action.body}
		case types.SET_IS_ERROR:
			return {...state, isError: action.body}
		default: 
			return state;
	}
}