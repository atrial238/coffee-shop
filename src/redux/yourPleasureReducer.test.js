import { yourPleasureReducer, getCoffeePleasure, types } from './yourPleasureReducer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureMockStore([thunk]);

const mockServiceCreator = (body, succeeds = true) => () =>
	new Promise((resolve, reject) => {
		setTimeout(() => (succeeds ? resolve(body) : reject('error')), 10);
});

describe('yourPleasureReducer', () => {
	const initState = {
		coffeePleasure: [],
		isLoading: false,
		isError: false
	}

	test('shoud return initial state', () => {
		const initialState = yourPleasureReducer(undefined, {});
		expect(initialState).toMatchSnapshot();
	});

	test('shoud set bestCoffee array', () => {
		const action =  {type: types.GET_PLEASURE_COFFEE, body: [{coffeeone: 'one'}, {coffeeTwo: 'two'}]};
		const state = yourPleasureReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('shoud set isLoading true', () => {
		const action =  {type: types.SET_IS_LOADING, body: true};
		const state = yourPleasureReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('shoud set isError true', () => {
		const action = {type: types.SET_IS_ERROR, body: true};
		const state = yourPleasureReducer(initState, action);
		expect(state).toMatchSnapshot();
	})
})

describe('getCoffeePleasure thunk', () => {
	let store;

	beforeEach(() => store = mockStore({}));
 
	it('dispatch action to store when getCoffeePleasure success', () => {
	
	const expectedActions = [
		{type: types.SET_IS_ERROR, body: false},
		{type: types.SET_IS_LOADING, body: true},
		{type: types.SET_IS_LOADING, body: false},
		{type: types.GET_PLEASURE_COFFEE, body: 'data'}
	]
	
	return store.dispatch(getCoffeePleasure(mockServiceCreator({coffeePleasure: 'data'})))
		.then(() => expect(store.getActions()).toEqual(expectedActions));
	});

	it('dispatch action to store when getCoffeePleasure failed', () => {
	
		const expectedActions = [
			{type: types.SET_IS_ERROR, body: false},
			{type: types.SET_IS_LOADING, body: true},
			{type: types.SET_IS_LOADING, body: false},
			{type: types.SET_IS_ERROR, body: true},
		]
		
		return store.dispatch(getCoffeePleasure(mockServiceCreator(null, false)))
			.then(() => expect(store.getActions()).toEqual(expectedActions))
			.catch(() => expect(store.getActions()).toEqual(expectedActions));
	});
 })