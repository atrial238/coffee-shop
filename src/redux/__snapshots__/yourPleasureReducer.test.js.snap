// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yourPleasureReducer shoud return initial state 1`] = `
Object {
  "coffeePleasure": Array [],
  "isError": false,
  "isLoading": false,
}
`;

exports[`yourPleasureReducer shoud set bestCoffee array 1`] = `
Object {
  "coffeePleasure": Array [
    Object {
      "coffeeone": "one",
    },
    Object {
      "coffeeTwo": "two",
    },
  ],
  "isError": false,
  "isLoading": false,
}
`;

exports[`yourPleasureReducer shoud set isError true 1`] = `
Object {
  "coffeePleasure": Array [],
  "isError": true,
  "isLoading": false,
}
`;

exports[`yourPleasureReducer shoud set isLoading true 1`] = `
Object {
  "coffeePleasure": Array [],
  "isError": false,
  "isLoading": true,
}
`;
